### PR TITLE
Fix crash throwing an incompletely constructed exception

### PR DESCRIPTION
**Pull request by jepler**
### Summary

This fixes my version of the reproducer for #17117, by preventing an incompletely constructed exception from actually being raised. 

### Testing

I added a test to the testsuite and it passed on the unix standard build.

### Trade-offs and Alternatives

In the comments on 17117, there's a link to a different fix in vm.c (to avoid trying to set the traceback info while unwinding) but I think avoiding throwing such an exception in the first place is preferable.

I tried placing the check in mp_obj_is_exception_instance instead, on the theory it might be better (catch more sites where such an object could escape). However, it wasn't a clear size win.

Since the reproducer I found involved throwing an exception from its own constructor, I considered catching exceptions that propagated from a constructor and ensuring `super().__init__` was called, similar to the non-error case. However, this would likely have been more code AND it wouldn't fix the case where the exception was thrown *and caught* within the constructor itself. Plus, we don't have the original reproducer for 17117 and maybe it found another way to do the same thing.

### CODE REVIEW

File: py/runtime.c [hunk 1]
```diff
@@ -1507,13 +1507,13 @@ mp_obj_t mp_make_raise_obj(mp_obj_t o) {
         o = mp_call_function_n_kw(o, 0, 0, NULL);
     }
 
-    if (mp_obj_is_exception_instance(o)) {
+    mp_obj_instance_t *obj = MP_OBJ_TO_PTR(o);
```
Comment (dpgeorge) [inline py/runtime.c]: This cast doesn't look legal, eg if `o` is not a concrete object, it could be a small int.

File: py/runtime.c [hunk 2]
```diff
@@ -1507,13 +1507,13 @@ mp_obj_t mp_make_raise_obj(mp_obj_t o) {
...
+    if (mp_obj_is_exception_instance(o) && obj->subobj[0] != &mp_native_base_init_wrapper_obj) {
```
Comment (dpgeorge) [inline py/runtime.c]: Do you think it's safe to dereference `obj->subobj[0]` even if `obj` may not be an `mp_obj_instance_t *`?  (The latter will only be true if `mp_obj_is_instance_type(o)`.)

File: tests/micropython/incomplete_exc.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class C(Exception):
+    def __init__(self):
+        raise self
+
+
+try:
+    C()
```
Comment (dpgeorge) [inline tests/micropython/incomplete_exc.py]: It looks like CPython allows the raise to occur... but I guess we don't need to match CPython for such a niche edge case.

File: py/runtime.c [hunk 4]
```diff
@@ -1507,13 +1507,17 @@ mp_obj_t mp_make_raise_obj(mp_obj_t o) {
         o = mp_call_function_n_kw(o, 0, 0, NULL);
     }
 
-    if (mp_obj_is_exception_instance(o)) {
-        // o is an instance of an exception, so use it as the exception
+    mp_obj_instance_t *obj = MP_OBJ_TO_PTR(o);
```
Comment (dpgeorge) [inline py/runtime.c]: It's a bit awkward defining this `obj` variable here when it's technically invalid if `o` is not a concrete object.

Instead I suggest just doing this cast inline in the expression below where `obj` is used.

File: py/runtime.c [hunk 5]
```diff
@@ -1507,13 +1507,17 @@ mp_obj_t mp_make_raise_obj(mp_obj_t o) {
...
+    if (mp_obj_is_native_exception_instance(o)) {
+        // o is an instance of a native, so use it as the exception
+        return o;
+    }
+    if (mp_obj_is_exception_instance(o) && obj->subobj[0] != MP_OBJ_FROM_PTR(&mp_native_base_init_wrapper_obj)) {
```
Comment (dpgeorge) [inline py/runtime.c]: There's an existing (static) function `get_native_exception()` which almost does exactly what you need.  So maybe we can save some code size doing the following:
```c
 if (mp_obj_is_exception_instance(o) && get_native_exception(o) != &mp_native_base_init_wrapper_obj) {
    return o;
} else {
    return mp_obj_new_exception_msg(...);
}
```

**Review by dpgeorge — COMMENTED** [review]
(I reviewed this ages ago but forgot to press "submit"!)

**Comment by jepler** [comment]
> (I reviewed this ages ago but forgot to press "submit"!)

I'm glad I'm not the only one this happens to.

I looked at various options and decided to try putting the check for `mp_native_base_init_wrapper_obj` in `mp_obj_is_exception_instance` because it does feel a bit cleaner and would apply the added check in places I may not have thought of; but it didn't save on code size. It also runs the check more often, needlessly, so runtime would become a bit slower.

The alternative of making `get_native_exception` non-static and calling it from `mp_make_raise_obj` was worse, though, +36 on bare-arm. I'm not really sure why but that's what my local size report run said.

```
    if (!mp_obj_is_exception_instance(o) || mp_obj_get_native_exception(o) == (void *)&mp_native_base_init_wrapper_obj) {
        // o cannot be used as an exception, so return a type error (which will be raised by the caller)
        return mp_obj_new_exception_msg(&mp_type_TypeError, MP_ERROR_TEXT("exceptions must derive from BaseException"));
    } else {
        return o;
    }
```

The best alternative was leaving the new check in mp_make_raise_obj but re-organizing it so that the `MP_OBJ_TO_PTR` call was after the `mp_obj_is_exception_instance`, which also I think addresses your concerns about using it on a non-boxed object.

File: py/runtime.c (hunk 2 above)
Comment (jepler) [reply py/runtime.c]: Oh, so for example an actual TypeError will not be "instance type". Probably not safe, then. mp_obj_exception_t is 16 bytes on a 32-bit platform (unix longlong) and `obj->subobj[0]` is at an offset of 16 bytes, so it addresses 4 bytes past the end of the object.

This would usually be fine (it'll hit memory where the bit value will not match the bit value of `mp_native_base_init_wrapper_obj`) but in fact if the object happened to be right at the end of a memory region it would segfault/hard fault. It would be hard to write a test to cover this specific situation.

**Comment by jepler** [comment]
Ah now that I understood about blindly accessing subobj, I think the least code growth alternative I've found is the current one, where the subobj check is added to `mp_obj_is_exception_instance` after the argument is determined to NOT be a native exception but to be of exception type, in which case we must have an instance object.

**Comment by dpgeorge** [comment]
I tested this PR with the two tests added in #18732.  Those tests don't crash which is great, although one of them has incorrect behaviour with this PR (compared to CPython).  But that seems to be the same incorrect behaviour as documented by the cpydiff added in this PR, so IMO it's acceptable.

I think we should merge this PR.

**Review by dpgeorge — APPROVED** [review]

**Comment by dpgeorge** [comment]
Rebased and merged in 20500558f8ca82e1d349a041f6fed8acbd6a2e3d through 2631b0614c16b7199aa814f26d04774da3d1443b

**Comment by jepler** [comment]
Thanks!
